apps/src/lib/levelbuilder/lesson-editor/ActivitySectionCard.jsx

```diff
@@ -0,0 +1,472 @@
+    newPosition: null,
+    startingPositions: null
+  };
+
+  handleDragStart = (position, {clientY}) => {
+    // The bounding boxes in this.metrics will be stale if the user scrolled the
+    // page since the last time this component was updated. Therefore, force the
+    // component to rerender so that this.metrics will be up to date.
+    this.forceUpdate(() => {
+      const startingPositions = this.props.activitySection.levels.map(level => {
+        const metrics = this.metrics[level.position];
+        return metrics.top + metrics.height / 2;
+      });
+      this.setState({
+        draggedLevelPos: position,
+        dragHeight: this.metrics[position].height + levelTokenMargin,
+        initialClientY: clientY,
+        newPosition: position,
+        startingPositions
+      });
+      window.addEventListener('selectstart', this.preventSelect);
+      window.addEventListener('mousemove', this.handleDrag);
+      window.addEventListener('mouseup', this.handleDragStop);
+    });
+  };
+
+  handleDrag = ({clientY}) => {
+    const delta = clientY - this.state.initialClientY;
+    const dragPosition = this.metrics[this.state.draggedLevelPos].top;
+    let newPosition = this.state.draggedLevelPos;
+    const currentPositions = this.state.startingPositions.map(
+      (midpoint, index) => {
+        const position = index + 1;
+        if (position === this.state.draggedLevelPos) {
+          return delta;
+        }
+        if (position < this.state.draggedLevelPos && dragPosition < midpoint) {
+          newPosition--;
+          return this.state.dragHeight;
+        }
+        if (
+          position > this.state.draggedLevelPos &&
+          dragPosition + this.state.dragHeight > midpoint
+        ) {
+          newPosition++;
+          return -this.state.dragHeight;
+        }
+        return 0;
+      }
+    );
+    this.setState({currentPositions, newPosition});
+    const targetActivitySectionPos = this.getTargetActivitySection(clientY);
+    this.props.setTargetActivitySection(targetActivitySectionPos);
+  };
+
+  // Given a clientY value of a location on the screen, find the ActivitySectionCard
+  // corresponding to that location, and return the position of the
+  // corresponding activity section within the script.
+  getTargetActivitySection = y => {
+    const {activitySectionMetrics} = this.props;
+    const activitySectionPos = Object.keys(activitySectionMetrics).find(
+      activitySectionPos => {
+        const activitySectionRect = activitySectionMetrics[activitySectionPos];
+        return (
+          y > activitySectionRect.top &&
+          y < activitySectionRect.top + activitySectionRect.height
+        );
+      }
+    );
+    return activitySectionPos ? Number(activitySectionPos) : null;
+  };
+
+  handleDragStop = () => {
+    const {
+      activitySection,
+      activityPosition,
+      targetActivitySectionPos
+    } = this.props;
+    if (targetActivitySectionPos === activitySection.position) {
```

Comment:
Sorry to keep you in suspense here Dani. What I discovered is that when I drag to reorder a script level, the left hand side of the check on this line is null. Digging in a bit, it looks like handleDrag is trying to store information in the state of the parent react component: https://github.com/code-dot-org/code-dot-org/blob/409f7f4c78ec2cfea4819c951ddd2cd5bc4999c5/apps/src/lib/levelbuilder/lesson-editor/ActivitySectionCard.jsx#L168-L169 but one problem is that those changes do not necessarily get back down into the ActivitySectionCard in time for the handleDragStop call to happen, because the ActivitySectionCard props only get updated when React's `render` method is called, which might not be right away.

Thoughts on how to proceed:
* just moving this into state within the ActivitySectionCard might not be enough, because setState calls do not happen immediately. there may be a way you can force them to be immediate but that might not be what you want since it could cause a re-render which could mess up dragging.
* ideally you would get your y coordinate from the handleDrag callback directly
* if that's not possible, you could try storing the clientY info on the react component outside of props/state. not ideal, but likely to be effective.

sorry if you knew most of this but wanted to get my thoughts down since I had spent some time looking at it.